### PR TITLE
[RLlib] Replace specs stub with real specs for base models

### PR DESCRIPTION
**Pull request by smorad**
Signed-off-by: Steven Morad <smorad@anyscale.com>

## Why are these changes needed?

With the introduction of #29027 and #29073, we can rebase #29043 and #29009. This swaps the `SpecsDict` stubs with the real thing for the base and torch models.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/models/temp_spec_classes.py [hunk 1]
```diff
@@ -4,48 +4,12 @@
 This file is to be removed once these modules are commited to master.
 """
 
-from typing import Any
+from ray.rllib.utils.nested_dict import NestedDict
```
Comment (kouroshHakha) [inline rllib/models/temp_spec_classes.py:7]: Let's remove this temp folder entirely, right? TensorDict is `NestedDict[Tensor]`. Just use that.

File: rllib/models/specs/specs_base.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 
 @DeveloperAPI
-class SpecsAbstract(abs.ABC):
```
Comment (kouroshHakha) [inline rllib/models/specs/specs_base.py:16]: already fixed on master

File: rllib/models/temp_spec_classes.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/models/temp_spec_classes.py:7]: You can define `TensorDict = NestedDict[Tensor]` in rllib.typing btw.

File: rllib/models/tests/test_base_model.py [hunk 3]
```diff
@@ -118,7 +120,7 @@ def test_initial_state(self):
         and outputs correct values."""
         output = NpRecurrentModelImpl().initial_state()
         desired = TensorDict({"in": np.arange(1)})
-        for k in output.flatten().keys() | desired.flatten().keys():
+        for k in output.shallow_keys() | desired.shallow_keys():
```
Comment (kouroshHakha) [inline rllib/models/tests/test_base_model.py:123]: You actually want the flattened keys which means you have to call `output.keys()`?

File: rllib/models/tests/test_torch_model.py [hunk 4]
```diff
@@ -106,10 +108,10 @@ def test_recurrent_unroll_and_filter(self):
         desired = TensorDict({"out": torch.arange(B * T * 3).reshape(B, T, 3)})
         desired_states = TensorDict({"out": torch.arange(B * 5).reshape(B, 5)})
 
-        for k in outputs.flatten().keys() | desired.flatten().keys():
+        for k in outputs.shallow_keys() | desired.shallow_keys():
```
Comment (kouroshHakha) [inline rllib/models/tests/test_torch_model.py:111]: same here

File: rllib/models/torch/model.py [hunk 5]
```diff
@@ -96,25 +96,25 @@ class TorchRecurrentModel(RecurrentModel, nn.Module, TorchModelIO):
         ...
         ...     @property
         ...     def input_spec(self):
-        ...         return SpecDict(
+        ...         return ModelSpecDict(
         ...             {"obs": "batch time hidden"}, hidden=self.config.input_size
```
Comment (kouroshHakha) [inline rllib/models/torch/model.py:100]: fix the docstring: `{"obs": TensorShape("b,t,h")}` etc (using `b,t,h` for conciseness.

**Review by kouroshHakha — APPROVED** [review]
Just a few nits/questions?

**Comment by kouroshHakha** [comment]
also plz merge master to make sure lint and other tests are also green.

File: rllib/models/temp_spec_classes.py (hunk 1 above)
Comment (smorad) [reply rllib/models/temp_spec_classes.py:7]: We still need the `ModelConfig` for now, I think.

**Comment by smorad** [comment]
I think most failing tests are unrelated. Lint error could be due to your/my PR?

```bash
!!! No API stability annotation found for:
--
  | ray.rllib.utils.nested_dict.StrKey
```

File: rllib/models/torch/model.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/models/torch/model.py]: dude you didn't do this :)

**Pull request closed without merging**